client/backoff/backoff.go

```diff
@@ -52,3 +53,29 @@ func (b *Backoff) Duration() time.Duration {
 func (b *Backoff) Reset() {
 	b.x = 0
 }
+
+// Retry calls a function until it succeeds or the context is done.
+// It will backoff if the function returns an error.
+// Once the context is done, retries will end and the most recent error will be returned.
+// Backoff is not reset by this function.
+func (b *Backoff) Retry(ctx context.Context, f func() error) error {
+	// If the context is already done, don't make any attempts to call f.
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	// Try calling f until it doesn't return an error or ctx is done.
+	for {
+		err := f()
+		if err == nil {
```

Comment:
ooo, sorry should've thought of this before, how about:
```go
for {
  err := f()
  if err != nil {
    select {
      …
    }
  }
  return nil
}
```
to stick with the indent-error-flow idiom?

(https://play.golang.org/p/r6Zc-WHqFS)